### PR TITLE
dffml: plugin: model: A pytorch IRIS classification model. [gsoc]

### PR DESCRIPTION
**Pull request by hsfzxjy**
I've re-implemented the tensorflow model using pytorch. Basic functionalities have been tested, but there may still remain some minor mistakes.

### CODE REVIEW

File: model/pytorch/setup.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    "scikit-learn"
+]
+
+setup(
+    name='dffml-model-pytorch',
+    version=version,
+    description='',
+    long_description=readme,
+    author='John Andersen',
+    author_email='john.s.andersen@intel.com',
```
Comment (pdxjohnny) [inline model/pytorch/setup.py]: Add your email ;)

**Comment by pdxjohnny** [comment]
Nice work!! Excited to see this progress.

**Comment by pdxjohnny** [comment]
Ping me when you've completed the test case so I know when to review.

**Comment by hsfzxjy** [comment]
@pdxjohnny I have added test cases and fixed multiple minor mistakes.

**Comment by pdxjohnny** [comment]
Great! Thank you! Please add to the he changelog too.

CHANGELOG.md

```
## [Unreleased]
### Added
- PyTorch model ... (add your explanation of what this is, one line)

## [0.1.2] - 2019-03-29
```

File: model/pytorch/setup.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        if line.startswith('VERSION'):
+            version = ast.literal_eval(line.strip().split('=')[-1].strip())
+            break
+
+with open(os.path.join(self_path, 'README.md'), 'r', encoding='utf-8') as f:
+    readme = f.read()
+
+INSTALL_REQUIRES = [
+    "torch",
+    "scikit-learn"
```
Comment (pdxjohnny) [inline model/pytorch/setup.py]: Please tag these with the latest versions e.g:

```python
INSTALL_REQUIRES = [
    "torch>=1.0.1.post2 ",
    "scikit-learn>=(look up on pypi)"
]
```

File: model/pytorch/setup.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+]
+
+setup(
+    name='dffml-model-pytorch',
+    version=version,
+    description='',
+    long_description=readme,
```
Comment (pdxjohnny) [inline model/pytorch/setup.py]: Ah, I forgot to update the `skel/` projects (#45)

Can you please add `long_description_content_type='text/markdown',` like in `model/tensorflow`:

https://github.com/intel/dffml/blob/c0ec62c4892b2c1a6ee0362861acd5e05ae717ab/model/tensorflow/setup.py#L27-L29

File: model/pytorch/tests/test_dfcn.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    def dtype(self) -> Type:
+        return int
+
+    def length(self) -> int:
+        return 1
+
+    async def calc(self, data: Data) -> int:
+        return 1 if data.src_url.lower().startswith('a') \
+            else 0
+
```
Comment (pdxjohnny) [inline model/pytorch/tests/test_dfcn.py]: nit: Remove double newline

File: model/pytorch/tests/test_dfcn.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+from typing import Type
+
+from dffml.repo import Repo, RepoData
+from dffml.source import Sources, RepoSource
+from dffml.feature import Data, Feature, Features
+from dffml.util.asynctestcase import AsyncTestCase
+
+from dffml_model_pytorch.model.dfcn import DFCN
+
+
```
Comment (pdxjohnny) [inline model/pytorch/tests/test_dfcn.py]: nit: Remove double newline

File: model/pytorch/tests/test_dfcn.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    async def calc(self, data: Data) -> int:
+        return 1 if data.src_url.lower().startswith('a') \
+            else 0
+
+
+class TestDFCN(AsyncTestCase):
+
+    @classmethod
+    def setUpClass(cls):
+
```
Comment (pdxjohnny) [inline model/pytorch/tests/test_dfcn.py]: nit: Remove newline

File: model/pytorch/tests/test_dfcn.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                          data={'features': {cls.feature.NAME: 1},
+                                'classification': 'a'}) for _ in range(0, 1000)]
+        cls.repos += [Repo('b' + str(random.random()),
+                           data={'features': {cls.feature.NAME: 0},
+                                 'classification': 'not a'}) for _ in range(0, 1000)]
+        cls.sources = Sources(RepoSource(*cls.repos))
+
+    @classmethod
+    def tearDownClass(cls):
+        # cls.model_dir.cleanup()
```
Comment (pdxjohnny) [inline model/pytorch/tests/test_dfcn.py]: I think you may have forgotten to uncomment this.

**Comment by pdxjohnny** [comment]
I just tested this against a dataset of mine, and realized that this is missing identification of which features this model will work on. Models should check the data types if features they are requested to train on and only select features they can operate on (e.g. this only works on ints and floats) See

https://github.com/intel/dffml/blob/bf3493e42315df0ec58a60c5d71d6cf28e622fca/model/tensorflow/dffml_model_tensorflow/model/dnn.py#L145-L169

For this model that probably means creating a `set()` of feature names which we've checked are of acceptable data types

**Comment by hsfzxjy** [comment]
@pdxjohnny Fixed it, and I added tests for such checking.

**Comment by pdxjohnny** [comment]
I'm sorry I don't have more canned test cases, but this doesn't work against features with arrays as data types (aka the git features), also it should check that the repos it is going to train on are of the requested classifications. This needs more tests before it can be merged.

**Comment by hsfzxjy** [comment]
It's OK. I'll later explore the whole project by myself and find out the necessary checking and test cases.

**Comment by pdxjohnny** [comment]
Awesome, thanks!

**Pull request closed without merging**